### PR TITLE
gh-102578: Optimise setting and deleting mutable attributes on non-dataclass subclasses of frozen dataclasses

### LINKED ISSUE #102578: [Enhancement] Speed up setting and deleting mutable attributes on non-dataclass subclasses of frozen dataclasses
# Feature or enhancement

The `dataclasses` library provides an easy way to create classes. The library will automatically generate relevant methods for the users.

Creating `dataclass`es with argument `frozen=True` will automatically generate methods `__setattr__` and `__delattr__` in `_frozen_get_del_attr`.

This issue proposes to change the `tuple`-based lookup to `set`-based lookup. Reduce the time complexity from $O(n)$ to $O(1)$.

```python
In [1]: # tuple-based

In [2]: %timeit 'a' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
9.91 ns ± 0.0982 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [3]: %timeit 'd' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
33.2 ns ± 0.701 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [4]: %timeit 'g' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
56.4 ns ± 0.818 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: # set-based

In [6]: %timeit 'a' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11.3 ns ± 0.0723 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [7]: %timeit 'd' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11 ns ± 0.106 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [8]: %timeit 'g' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11.1 ns ± 0.126 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```

A tiny benchmark script:

```python
from contextlib import suppress
from dataclasses import FrozenInstanceError, dataclass

@dataclass(frozen=True)
class Foo2:
    a: int
    b: int

foo2 = Foo2(1, 2)

def bench2(inst):
    with suppress(FrozenInstanceError):
        inst.a = 0
    with suppress(FrozenInstanceError):
        inst.b = 0

@dataclass(frozen=True)
class Foo7:
    a: int
    b: int
    c: int
    d: int
    e: int
    f: int
    g: int

foo7 = Foo7(1, 2, 3, 4, 5, 6, 7)

def bench7(inst):
    with suppress(FrozenInstanceError):
        inst.a = 0
    with suppress(FrozenInstanceError):
        inst.b = 0
    with suppress(FrozenInstanceError):
        inst.c = 0
    with suppress(FrozenInstanceError):
        inst.d = 0
    with suppress(FrozenInstanceError):
        inst.e = 0
    with suppress(FrozenInstanceError):
        inst.f = 0
    with suppress(FrozenInstanceError):
        inst.g = 0

class Bar(Foo7):
    def __init__(self, a, b, c, d, e, f, g):
        super().__init__(a, b, c, d, e, f, g)
        self.baz = 0

def bench(inst):
    inst.baz = 1
```

Result:

`set`-based lookup:

```python
In [2]: %timeit bench2(foo2)
1.08 µs ± 28.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.81 µs ± 20.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit bench(bar)
249 ns ± 6.31 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

`tuple`-based lookup (original):

```python
In [2]: %timeit bench2(foo2)
1.15 µs ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.97 µs ± 15.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit bench(bar)
269 ns ± 4.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
```

Result:

`set`-based lookup:

```python
In [2]: %timeit bench2(foo2)
1.08 µs ± 28.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.81 µs ± 20.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

`tuple`-based lookup (original):

```python
In [2]: %timeit bench2(foo2)
1.15 µs ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.97 µs ± 15.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

The `set`-based is constantly faster than the old approach. And the theoretical time complexity is also smaller ($O(1)$ vs. $O(n)$).

Ref: #102573

# Pitch

(Explain why this feature or enhancement should be implemented and how it would be used.
 Add examples, if applicable.)

In the autogenerate `__setattr__` and `__delattr__`, they have a sanity check at the beginning of the method. For example:

```python
def __setattr__(self, name, value):
    if type(self) is {{UserType}} or name in ({{a tuple of field names}}):
        raise FrozenInstanceError(f"cannot assign to field {name!r}")
    super(cls, self).__setattr__(name, value)
```

If someone inherits the frozen dataclass, the sanity check will take $O(n)$ time on the `tuple__contains__(...)` and finally calls `super().__setattr__(...)`. For example:

```python
@dataclass(frozen=True)
class FrozenBase:
    x: int
    y: int
    ... # N_FIELDS

class Foo(FrozenBase):
    def __init__(self, x, y, somevalue, someothervalue):
        super().__init__(x, y)
        self.somevalue = somevalue            # takes O(N_FIELDS)
        self.someothervalue = someothervalue  # takes O(N_FIELDS) time again

foo = Foo(1, 2, 3, 4)
foo.extravalue = extravalue  # takes O(N_FIELDS) time again
```

# Previous discussion

N/A.

### Linked PRs
* gh-102573

### PR DESCRIPTION
**Pull request by XuehaiPan**
Creating `dataclass`es with argument `frozen=True` will automatically generate methods `__setattr__` and `__delattr__` in `_frozen_get_del_attr`.

This PR changes `tuple`-based lookup to `set`-based lookup. Reduce the time complexity from $O(n)$ to $O(1)$.

```python
In [1]: # tuple-based

In [2]: %timeit 'a' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
9.91 ns ± 0.0982 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [3]: %timeit 'd' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
33.2 ns ± 0.701 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [4]: %timeit 'g' in ('a', 'b', 'c', 'd', 'e', 'f', 'g')
56.4 ns ± 0.818 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: # set-based

In [6]: %timeit 'a' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11.3 ns ± 0.0723 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [7]: %timeit 'd' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11 ns ± 0.106 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)

In [8]: %timeit 'g' in {'a', 'b', 'c', 'd', 'e', 'f', 'g'}
11.1 ns ± 0.126 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```

A tiny benchmark script:

```python
from contextlib import suppress
from dataclasses import FrozenInstanceError, dataclass

@dataclass(frozen=True)
class Foo2:
    a: int
    b: int

foo2 = Foo2(1, 2)

def bench2(inst):
    with suppress(FrozenInstanceError):
        inst.a = 0
    with suppress(FrozenInstanceError):
        inst.b = 0

@dataclass(frozen=True)
class Foo7:
    a: int
    b: int
    c: int
    d: int
    e: int
    f: int
    g: int

foo7 = Foo7(1, 2, 3, 4, 5, 6, 7)

def bench7(inst):
    with suppress(FrozenInstanceError):
        inst.a = 0
    with suppress(FrozenInstanceError):
        inst.b = 0
    with suppress(FrozenInstanceError):
        inst.c = 0
    with suppress(FrozenInstanceError):
        inst.d = 0
    with suppress(FrozenInstanceError):
        inst.e = 0
    with suppress(FrozenInstanceError):
        inst.f = 0
    with suppress(FrozenInstanceError):
        inst.g = 0

class Bar(Foo7):
    def __init__(self, a, b, c, d, e, f, g):
        super().__init__(a, b, c, d, e, f, g)
        self.baz = 0

def bench(inst):
    inst.baz = 1
```

Result:

`set`-based lookup:

```python
In [2]: %timeit bench2(foo2)
1.08 µs ± 28.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.81 µs ± 20.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit bench(bar)
249 ns ± 6.31 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

`tuple`-based lookup (original):

```python
In [2]: %timeit bench2(foo2)
1.15 µs ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit bench7(foo7)
3.97 µs ± 15.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [4]: %timeit bench(bar)
269 ns ± 4.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

The `set`-based is constantly faster than the old approach. And the theoretical time complexity is also smaller ($O(1)$ vs. $O(n)$).

* Issue: gh-102578

Resolves #102578

### CODE REVIEW

File: Lib/dataclasses.py
Comment (TeamSpen210) [inline Lib/dataclasses.py]: This call wouldn’t be constant-foldable, so it’s probably a regression. Since `field_str` is being formatted into the one line twice below, if that was moved up here we could instead just eliminate the second clause entirely in this case.

Comment (XuehaiPan) [reply Lib/dataclasses.py]: Make sense. Thanks for the advice.

File: Misc/NEWS.d/next/Library/2023-03-10-10-10-00.gh-issue-102573.GQ7sV6.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-10-10-10-00.gh-issue-102573.GQ7sV6.rst]: This news entry only describes the changes you have made to the internal implementation, but users of frozen dataclasses will probably not care much about that kind of thing. You should change the news entry so that it describes the impact that this change will have on users of frozen dataclasses.

E.g. does this speed up the time it takes to set certain attributes? Does it speed up the time taken to construct frozen dataclass instances? If it does speed certain things up, how much does it speed things up by?

**Comment by AlexWaygood** [comment]
The `issue-number` check is failing because `gh-102573` does not yet exist as a GitHub issue :)

You should create an issue to describe the change you're proposing here, and then link to it in this PR title.

File: Misc/NEWS.d/next/Library/2023-03-10-10-10-00.gh-issue-102573.GQ7sV6.rst
Comment (XuehaiPan) [reply Misc/NEWS.d/next/Library/2023-03-10-10-10-00.gh-issue-102573.GQ7sV6.rst]: I update the news entry with an appropriate description.

**Comment by XuehaiPan** [comment]
Opened a linked issue and added some benchmark results.

File: Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst]: This is better, but still doesn't explain exactly what is being sped up here (users are very unlikely to be calling `__setattr__` or `__delattr__` directly, but they might be trying to set or delete attributes -- so we should talk about speeding up setting/deleting attributes rather than speeding up `__setattr__` or `__delattr__`).

My question when reading this is, "Well, why should I care about how fast `__setattr__` and `__delattr__` are on frozen dataclasses, given that `__setattr__` and `__delattr__` always raise exceptions on frozen dataclasses? The benchmark you've given on the issue also doesn't answer this question -- it seems like it consists of very contrived idioms that users would be unlikely to use in real-world code.

If the answer is "not much important is being sped up here, actually", then we need to ask whether it's worth making this change at all :)

Comment (XuehaiPan) [reply Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst]: Sorry for the unclear description. I updated the news entry and the Pitch section in the linked issue.

> My question when reading this is, "Well, why should I care about how fast `__setattr__` and `__delattr__` are on frozen dataclasses, given that `__setattr__` and `__delattr__` always raise exceptions on frozen dataclasses?

My answer is no. It will not always raise exceptions. I added an example in the news entry and the Pitch section in the linked issue.

An example use case for inheriting frozen `dataclass`es.

```python
@dataclass(frozen=True)
class FrozenBase:
    x: int
    y: int
    ... # N_FIELDS

class Foo(FrozenBase):
    def __init__(self, x, y, somevalue, someothervalue):
        super().__init__(x, y)
        self.somevalue = somevalue            # takes O(N_FIELDS)
        self.someothervalue = someothervalue  # takes O(N_FIELDS) time again

foo = Foo(1, 2, 3, 4)
foo.extravalue = extravalue  # takes O(N_FIELDS) time again
```

File: Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst
Comment (AlexWaygood) [inline Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst]: Thank you! I now understand perfectly what this speeds up, and agree that it seems like a good idea. But we can be a lot more concise for this news entry :)

```suggestion
Speed up setting or deleting mutable attributes on non-dataclass subclasses of
frozen dataclasses. Due to the implementation of ``__setattr__`` and
``__delattr__`` for frozen dataclasses, this previously had a time complexity
of ``O(n)``. It now has a time complexity of ``O(1)``.
```

File: Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst
Comment (AlexWaygood) [reply Misc/NEWS.d/next/Library/2023-03-10-13-21-16.gh-issue-102578.-gujoI.rst]: Thank you! The subclass case seems like a valid use case. I now agree that this seems like a good idea.

**Comment by AlexWaygood** [comment]
Looks like the CLA check has started failing with the latest commit -- if you're using two email address, you may have to sign it with both email addresses, unfortunately :(

**Comment by XuehaiPan** [comment]
Seems that the CLA check has passed now.

**Review by AlexWaygood — APPROVED** [review]
Looks great to me. I verified the speedup by running this benchmark locally (which is a little different to the one you posted in your issue and the one you posted in this PR).

<details>
<summary>Benchmark:</summary>

```python
import dataclasses
import string
import time

Foo = dataclasses.make_dataclass(
    "Foo",
    [(letter, int) for letter in string.ascii_lowercase],
    frozen=True
)

class Bar(Foo): ...

instance = Bar(*range(26))
t0 = time.perf_counter()
for _ in range(10_000_000):
    instance.foo = 1
    del instance.foo
print(f"{time.perf_counter() - t0:.2f}")
```

</details>

The result of this benchmark is 15.15 seconds on `main` on my machine (`--pgo` non-debug build), but 6.44 seconds with this patch applied.

I'll wait for a thumbs-up from @ericvsmith, @carljm, or another core dev before merging, but this has my approval -- thanks!

File: Lib/dataclasses.py
Comment (ericvsmith) [inline Lib/dataclasses.py]: Not that it matters much, but the zero length case could just avoid the `or name in ...` test entirely. Maybe fields_str should become fields_test, and then set it to `or name in {<<generated set literal>>}` or set it to an empty string if there are no fields. Then change the generated code to `f'if type(self) is cls {fields_test}:'` Although that doesn't read very well. Maybe tweak fields_test to be something else.

This could be part of a different PR, or include it here. But in any event I'm not positive that the zero length case actually has a test. We should make sure it does for this PR.

File: Lib/dataclasses.py
Comment (AlexWaygood) [inline Lib/dataclasses.py]: ?

```suggestion
```

File: Lib/dataclasses.py
Comment (XuehaiPan) [reply Lib/dataclasses.py]: > But in any event I'm not positive that the zero length case actually has a test. We should make sure it does for this PR.

Added a small test for empty frozen dataclass.

File: Lib/test/test_dataclasses.py
Comment (carljm) [inline Lib/test/test_dataclasses.py]: I think we can leave out these two lines. If we need a test that `FrozenInstanceError` is a subclass of `AttributeError`, `test_frozen_empty` is an odd place to put that test. It doesn't seem related to the main purpose of the test.

File: Lib/test/test_dataclasses.py
Comment (carljm) [inline Lib/test/test_dataclasses.py:2778]: Can we also test that you _can_ add an attribute to a non-dataclass subclass of the empty frozen dataclass?

**Review by carljm — CHANGES_REQUESTED** [review]
Code changes LGTM. A couple comments on the new test.

**Comment by XuehaiPan** [comment]
I have made the requested changes; please review again

File: Lib/test/test_dataclasses.py
Comment (XuehaiPan) [inline Lib/test/test_dataclasses.py]: Should we also check the raised exception is exactly `AttributeError` rather than `FrozenInstanceError` (subclass of `AttributeError`)? This exception is raised from `super().__delattr__(...)`.

**Review by carljm — APPROVED** [review]
Looks good to me; thanks for the perf improvement!

File: Lib/test/test_dataclasses.py
Comment (AlexWaygood) [reply Lib/test/test_dataclasses.py]: That would get you extra brownie points :)

I think you could do it like this:

```suggestion
        with self.assertRaises(AttributeError) as cm:
            del s.cached
        
        self.assertNotIsInstance(cm.exception, FrozenInstanceError)
```

Comment (XuehaiPan) [reply Lib/test/test_dataclasses.py]: Thanks for the guidance. Added.

**Review by AlexWaygood — APPROVED** [review]
Looks great!

**Pull request merged**
